creusot-contracts/src/ghost/invariant.rs

```diff
@@ -190,35 +198,99 @@ impl View for Tokens<'_> {
     }
 }
 
-/// A ghost structure, that holds a piece of data (`T`) together with an
-/// [protocol](Protocol).
-#[opaque]
-pub struct LocalInvariant<T: Protocol> {
-    value: UnsafeCell<T>,
-}
-
-/// A variant of [`Invariant`] for use in [`LocalInvariant`]s.
+/// A variant of [`Invariant`] for use in [`AtomicInvariant`]s and [`NonAtomicInvariant`]s.
 ///
-/// This allows to specify an invariant that depends on some [public data](LocalInvariant::public).
+/// This allows to specify an invariant that depends on some [public data](NonAtomicInvariant::public).
 pub trait Protocol {
     type Public;
 
     #[logic(prophetic)]
     fn protocol(self, data: Self::Public) -> bool;
 }
 
-/// Define method call syntax for [`LocalInvariant::open`].
-pub trait LocalInvariantExt<'a> {
+#[opaque]
+pub struct AtomicInvariant<T: Protocol>(PhantomData<T>);
+
+unsafe impl<T: Protocol + Send> Sync for AtomicInvariant<T> {}
+
+impl<T: Protocol> AtomicInvariant<T> {
+    /// Construct a `AtomicInvariant`
+    ///
+    /// # Parameters
+    /// - `value`: the actual data contained in the invariant. Use [`Self::open`] to
+    /// access it. Also called the 'private' part of the invariant.
+    /// - `public`: the 'public' part of the invariant.
+    /// - `namespace`: the namespace of the invariant.
+    ///   This is required to avoid [open](Self::open)ing the same invariant twice.
+    #[trusted]
+    #[requires(value.protocol(*public))]
+    #[ensures(result.public() == *public)]
+    #[ensures(result.namespace() == *namespace)]
+    #[check(ghost)]
+    pub fn new(
+        value: Ghost<T>,
+        public: Snapshot<T::Public>,
+        namespace: Snapshot<Namespace>,
+    ) -> Ghost<Self> {
+        Ghost::conjure()
+    }
+
+    /// Get the namespace associated with this invariant.
+    #[logic(opaque)]
+    pub fn namespace(self) -> Namespace {
+        dead
+    }
+
+    /// Get the 'public' part of this invariant.
+    #[logic(opaque)]
+    pub fn public(self) -> T::Public {
+        dead
+    }
+
+    /// Open the invariant to get the data stored inside.
+    ///
+    /// This will call the closure `f` with the inner data. You must restore the
+    /// contained [`Protocol`] before returning from the closure.
```

Comment:
```suggestion
    /// contained [`Protocol`] before returning from the closure.
    
    /// Note: this function can only be called from ghost code, because atomic invariants are always wrapped in `Ghost`. This guarantees atomicity.
```